# Data Understanding 
<!-- # E-Commerce Marketplace Analytics -->

### 1. Project Objective

The objective of this project is to analyze an e-commerce marketplace using the Olist Brazilian E-Commerce dataset.

The analysis focuses on sales, customers, products, sellers, payments, delivery performance, and customer reviews using SQL, Python, and Power BI.

### 2. Importing the Dataset

The Olist dataset contains multiple CSV files. Each CSV represents a different part of the e-commerce business.

First, we will import the required Python libraries and load each CSV file into a separate Pandas DataFrame.

In [1]:
import pandas as pd 

In [2]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")

order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")

payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")

reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")

products = pd.read_csv("../data/raw/olist_products_dataset.csv")

sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")

geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")

category_translation = pd.read_csv("../data/raw/product_category_name_translation.csv")


### 3. Understanding the Dataset

We will first check the structure of each table using `info()`.

This helps us understand the number of rows, columns, data types, and missing values.

In [3]:
orders.info()
order_items.info()
payments.info()
reviews.info()
products.info()
sellers.info()
geolocation.info()
category_translation.info()
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB
<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 no

### 4. Dataset Size

We will check the number of rows and columns in each table using the `shape` property.

In [4]:
print("Customers:", customers.shape)
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Payments:", payments.shape)
print("Reviews:", reviews.shape)
print("Products:", products.shape)
print("Sellers:", sellers.shape)
print("Geolocation:", geolocation.shape)
print("Category Translation:", category_translation.shape)

Customers: (99441, 5)
Orders: (99441, 8)
Order Items: (112650, 7)
Payments: (103886, 5)
Reviews: (99224, 7)
Products: (32951, 9)
Sellers: (3095, 4)
Geolocation: (1000163, 5)
Category Translation: (71, 2)


### 5. Previewing the Tables

The `head()` function displays the first five rows of a DataFrame.

This helps us understand what type of information each table contains.

In [5]:
len(orders)
category_translation.head()
sellers.head()
products.head()
orders.head()
reviews.head()
order_items.head()
payments.head()
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


---

### 6. Understanding the Orders Table

The Orders table contains information about each order placed on the platform.

Important columns include:

- `order_id` — unique identifier of an order
- `customer_id` — customer associated with the order
- `order_status` — current status of the order
- `Order timestamps` — information about the order lifecycle



In [6]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

#### Checking Unique Orders

We will compare the total number of rows with the number of unique `order_id` values to understand whether each row represents a unique order.

In [7]:
print('Rows:',len(orders))
print('Unique Order IDs:', orders['order_id'].nunique())
print('Unique Customer IDs:', orders['customer_id'].nunique())

Rows: 99441
Unique Order IDs: 99441


Unique Customer IDs: 99441


---

### 7. Understanding the Customers Table

The Customers table contains customer information such as customer IDs, location, city, and state.

The dataset contains both `customer_id` and `customer_unique_id`.

`customer_unique_id` is useful for identifying distinct customers across the marketplace.

In [8]:
print(customers.columns)

len(customers)
print('Rows:',len(customers))
print('unique customer IDs:', customers['customer_id'].nunique())
print('unique "unique customer IDs" :', customers['customer_unique_id'].nunique())


Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')
Rows: 99441
unique customer IDs: 99441
unique "unique customer IDs" : 96096


### Identifying Repeat Customers

A customer can appear more than once through `customer_unique_id`.

We will count how many times each unique customer appears and identify customers with more than one record.

In [9]:
customer_frequency= customers['customer_unique_id'].value_counts()

repeat_customers= customer_frequency[customer_frequency >1]
print(repeat_customers)
print('Total Repeat Customers :',len(repeat_customers))

print('Repeat Customer %age :' , round(len(repeat_customers)/customers['customer_unique_id'].nunique()*100,2))

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
                                    ..
574e36a816bf7a295404e686062fc0de     2
76f7d40948174f5cde0fcf8e6147b59d     2
88fcb680dc8cd2e540010c0750b78242     2
2dcbfb99d95c7ba6eb992a98627f45b8     2
da29b8a3970adddb9ed2dc8a492065af     2
Name: count, Length: 2997, dtype: int64
Total Repeat Customers : 2997
Repeat Customer %age : 3.12


---

### 8. Understanding the Order Items Table

The `order_items` table contains information about individual products included in customer orders.

The main columns are `order_id`, `product_id`, `seller_id`, `price`, and `freight_value`. An order can contain multiple order items, allowing us to analyze product and seller performance.

In [10]:
order_items.columns
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


### Orders vs Order Items

An order can contain multiple products. Therefore, the number of order-item records can be greater than the number of orders.

In [11]:
print('Total Orders :', len(orders))
print('Total Order Items :', len(order_items))


# unique orders in order_items

print('unique orders in order_items :',order_items['order_id'].nunique())

Total Orders : 99441
Total Order Items : 112650
unique orders in order_items : 98666


The difference between the number of orders and order-item records shows that some orders contain multiple items.

The `order_items` table can therefore be used to analyze product and seller performance.

### An order item represents WHAT?

In [12]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [13]:
order_items["order_item_id"].value_counts().sort_index()

order_item_id
1     98666
2      9803
3      2287
4       965
5       460
6       256
7        58
8        36
9        28
10       25
11       17
12       13
13        8
14        7
15        5
16        3
17        3
18        3
19        3
20        3
21        1
Name: count, dtype: int64

### 9. Main Table Relationships

The main relationships in the dataset are:

- Customers → Orders using `customer_id`
- Orders → Order Items using `order_id`
- Orders → Payments using `order_id`
- Orders → Reviews using `order_id`
- Order Items → Products using `product_id`
- Order Items → Sellers using `seller_id`
- Products → Category Translation using the product category name

The Orders table is the central table connecting the customer's order with payments, reviews, and purchased items.

## 10. Initial Findings

The initial exploration shows that:

- The dataset contains 99,441 orders.
- There are 96,096 unique customers based on `customer_unique_id`.
- 2,997 customers have more than one customer record.
- The order_items table contains 112,650 records.
- One order can contain multiple order items.
- `customer_unique_id` will be used for customer-level analysis.
- `order_id`, `customer_id`, `product_id`, and `seller_id` are important identifiers for connecting the tables.

---

### Payments Table

The Payments table contains information about how customers paid for their orders.

Important columns include:

- `order_id` — identifies the order
- `payment_type` — payment method used
- `payment_installments` — number of installments
- `payment_value` — amount paid

This table will later be used to analyze payment methods, payment amounts, and installment behavior.

## 11. Conclusion

The Olist dataset is a relational e-commerce dataset containing information about customers, orders, products, sellers, payments, reviews, and logistics.

The basic structure and relationships between the tables have been understood. The next phase will focus on checking data quality, including missing values, duplicates, and data types.

In [14]:
print(payments.head())
print(payments.columns)

                           order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  
Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')


---

### Reviews Table

The Reviews table contains customer feedback about their orders.

Important columns include:

- `order_id` — identifies the order
- `review_score` — customer rating from 1 to 5
- `review_comment_title` — review title
- `review_comment_message` — review text

This table will later be used to analyze customer satisfaction and review scores.

In [15]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [16]:
reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

---

### Products Table

The Products table contains information about the products sold on the marketplace.

Important columns include:

- `product_id` — identifies the product
- `product_category_name` — product category
- Product dimensions and weight — physical product information

This table will later be used to analyze product and category performance.

In [17]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [18]:
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='str')

---

### Sellers Table

The Sellers table contains information about sellers operating on the marketplace.

Important columns include:

- `seller_id` — identifies the seller
- `seller_zip_code_prefix` — seller location
- `seller_city` — seller city
- `seller_state` — seller state

This table will later be used to analyze seller performance and geographic distribution.

In [19]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [20]:
sellers.columns

Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='str')

---

### Geolocation Table

The Geolocation table contains geographic information associated with Brazilian ZIP code prefixes.

It can be used later for geographic and location-based analysis.

In [21]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [22]:
geolocation.columns

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='str')

---

### Category Translation Table

The Category Translation table maps Portuguese product category names to their English translations.

This table will help make product category analysis easier to understand.

In [23]:
category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [24]:
category_translation.columns

Index(['product_category_name', 'product_category_name_english'], dtype='str')

## Phase 1 Summary

The Olist dataset contains multiple related tables covering customers, orders, products, sellers, payments, reviews, category translation and geographic information.

The `orders` table represents the main order lifecycle, while `order_items` connects orders with products and sellers. Payments and reviews provide information about transactions and customer satisfaction.

The dataset structure is now understood and can be used for the next stage of the project: data quality analysis.

---